src/middleware/_utils.ts

```diff
@@ -1,30 +1,46 @@
 import type { IncomingMessage, OutgoingMessage } from "node:http";
 import type { ProxyServer } from "../server";
-import type { ProxyServerOptions } from "../types";
+import type {
+  NormalizedProxyTarget,
+  NormalizedProxyTargetUrl,
+  ProxyServerOptions,
+} from "../types";
+import type { Socket } from "node:net";
 
-export type ProxyMiddleware = (
+export type ResOfType<T extends "web" | "ws"> = T extends "ws"
+  ? T extends "web"
+    ? OutgoingMessage | Socket
+    : Socket
+  : T extends "web"
+    ? OutgoingMessage
+    : never;
+
+export type ProxyMiddleware<T extends OutgoingMessage | Socket> = (
   req: IncomingMessage,
-  res: OutgoingMessage,
-  opts: ProxyServerOptions & { target: URL; forward: URL },
-  server?: ProxyServer,
+  res: T,
+  opts: ProxyServerOptions & {
+    target: NormalizedProxyTarget;
+    forward: NormalizedProxyTargetUrl;
+  },
+  server: ProxyServer,
   head?: Buffer,
-  callback?: (
-    err: any,
-    req: IncomingMessage,
-    socket: OutgoingMessage,
-    url?: any,
-  ) => void,
+  callback?: (err: any, req: IncomingMessage, socket: T, url?: any) => void,
 ) => void | true;
 
-export function defineProxyMiddleware(m: ProxyMiddleware) {
+export function defineProxyMiddleware<
+  T extends OutgoingMessage | Socket = OutgoingMessage,
+>(m: ProxyMiddleware<T>) {
   return m;
 }
 
 export type ProxyOutgoingMiddleware = (
   req: IncomingMessage,
   res: OutgoingMessage,
   proxyRes: IncomingMessage,
-  opts: ProxyServerOptions & { target: URL; forward: URL },
+  opts: ProxyServerOptions & {
+    target: NormalizedProxyTarget;
+    forward: NormalizedProxyTargetUrl;
+  },
```

Comment:
This was wrong.
https://github.com/unjs/httpxy/blob/ea01d8c553f18efaa69c4df70f3173c202d06460/src/server.ts#L153-L155
`string` is normalized to `URL`, but the `ProxyTargetDetailed` type is passed as-is, so the type should be `URL | ProxyTargetDetailed` for `target`.